### PR TITLE
R4R : Move asset to token submodule

### PR DESCRIPTION
**Pull request by zhiqiang-bianjie**
Resolves : #2141

### CODE REVIEW

File: modules/asset/module.go [hunk 1]
```diff
@@ -29,6 +29,8 @@ var (
 // AppModuleBasic app module basics object
 type AppModuleBasic struct{}
 
+var _ module.AppModuleBasic = AppModuleBasic{}
```
Comment (SegueII) [inline modules/asset/module.go]: redundant

File: modules/asset/module.go [hunk 2]
```diff
@@ -47,7 +49,8 @@ func (AppModuleBasic) DefaultGenesis() json.RawMessage {
 // ValidateGenesis module validate genesis
 func (AppModuleBasic) ValidateGenesis(bz json.RawMessage) error {
 	var data GenesisState
-	if err := ModuleCdc.UnmarshalJSON(bz, &data); err != nil {
+	err := ModuleCdc.UnmarshalJSON(bz, &data)
+	if err != nil {
```
Comment (SegueII) [inline modules/asset/module.go]: merge into one line

File: modules/asset/genesis.go [hunk 3]
```diff
@@ -1,35 +1,47 @@
...
+
+// ValidateGenesis validates the provided asset genesis state to ensure the
+// expected invariants holds.
+func ValidateGenesis(data GenesisState) error {
+	// validate tokens
+	if err := token.ValidateGenesis(data.TokenState); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (SegueII) [inline modules/asset/genesis.go]: return token.ValidateGenesis(data.TokenState)

File: modules/asset/01-token/internal/types/msgs.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+	}
+
+	if err := ValidateSupply(msg.InitialSupply, msg.MaxSupply); err != nil {
+		return err
+	}
+
+	if err := ValidateScale(msg.Scale); err != nil {
+		return err
+	}
+	return nil
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/msgs.go]: could return last check directly

File: modules/asset/01-token/internal/types/msgs.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+	if msg.SrcOwner.Equals(msg.DstOwner) {
+		return ErrInvalidToAddress(DefaultCodespace, fmt.Sprintf("the new owner must not be same as the original owner"))
+	}
+
+	// check the Symbol
+	if err := ValidateSymbol(msg.Symbol); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/msgs.go]: ditto

File: modules/asset/01-token/internal/types/msgs.go [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+
+	if err := ValidateName(msg.Name); DoNotModify != msg.Name && err != nil {
+		return err
+	}
+
+	if err := ValidateSymbol(msg.Symbol); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/msgs.go]: ditto

File: modules/asset/01-token/internal/types/genesis.go [hunk 7]
```diff
@@ -16,14 +30,21 @@ func NewGenesisState(params Params, tokens Tokens) GenesisState {
...
 // ValidateGenesis validates the provided asset genesis state to ensure the
 // expected invariants holds.
 func ValidateGenesis(data GenesisState) error {
-	if err := data.Params.Validate(); err != nil {
-		return err
+	//if err := data.Params.Validate(); err != nil {
+	//	return err
+	//}
+	//return data.Tokens.Validate()
+	return data.Params.Validate()
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/genesis.go]: check  `data.Params.Validate()` & `data.Params.Validate()`
implement `Params.Validate()`

File: modules/asset/handler.go [hunk 8]
```diff
@@ -1,147 +1,29 @@
...
 
 import (
-	"fmt"
-
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	token "github.com/irisnet/irishub/modules/asset/01-token"
 )
 
-// NewHandler returns a handler for all "asset" type messages.
+// handle all "asset" type messages.
```
Comment (chengwenxi) [inline modules/asset/handler.go]: comment should start with the func name.

File: modules/asset/01-token/internal/types/genesis.go [hunk 9]
```diff
@@ -36,10 +36,14 @@ func DefaultGenesisState() GenesisState {
...
-	//	return err
-	//}
-	//return data.Tokens.Validate()
+	for _, token := range data.Tokens {
+		if err := ValidateName(token.Name); err != nil {
+			return err
+		}
+		if err := ValidateScale(token.Scale); err != nil {
+			return err
+		}
```
Comment (chengwenxi) [inline modules/asset/01-token/internal/types/genesis.go]: check min-unit?

File: modules/asset/01-token/client/rest/rest.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	BaseTx        rest.BaseReq   `json:"base_tx" yaml:"base_tx"`
+	Owner         sdk.AccAddress `json:"owner" yaml:"owner"` //  Owner of the token
+	Symbol        string         `json:"symbol" yaml:"symbol"`
+	Name          string         `json:"name" yaml:"name"`
+	Scale         uint8          `json:"scale" yaml:"scale"`
+	MinUnit       string         `json:"min_unit" yaml:"min_unit"`
+	InitialSupply uint64         `json:"initial_supply" yaml:"initial_supply"`
+	MaxSupply     uint64         `json:"max_supply" yaml:"max_supply"`
+	Mintable      bool           `json:"mintable" yaml:"mintable"`
+}
```
Comment (SegueII) [inline modules/asset/01-token/client/rest/rest.go]: add attribute comments for each struct

File: modules/asset/01-token/internal/keeper/ante.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+func NewValidateTokenFeeDecorator(tk Keeper, ak types.AccountKeeper) ValidateTokenFeeDecorator {
+	return ValidateTokenFeeDecorator{
+		tk: tk,
+		ak: ak,
+	}
+}
+
+func (dtf ValidateTokenFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
```
Comment (SegueII) [inline modules/asset/01-token/internal/keeper/ante.go]: comments

File: modules/asset/01-token/internal/keeper/querier.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/irisnet/irishub/modules/asset/01-token/internal/types"
+)
+
+func QuerierToken(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) ([]byte, sdk.Error) {
```
Comment (SegueII) [inline modules/asset/01-token/internal/keeper/querier.go]: comments for `QuerierToken` `QuerierTokens` `QuerierFees` `QuerierParameters `

File: modules/asset/01-token/internal/types/keys.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+	// DefaultParamspace default name for parameter store
+	DefaultParamspace = SubModuleName
+)
+
+func KeyPath(prefix int) []byte {
+	return types.KeyPrefixBytes(SubModuleName, prefix)
+}
+
+func KeyTokenPrefix() []byte {
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/keys.go]: comments

File: modules/asset/01-token/internal/types/expected_keepers.go [hunk 14]
```diff
@@ -2,20 +2,25 @@ package types
...
 	GetModuleAddress(name string) sdk.AccAddress
-	GetModuleAccount(ctx sdk.Context, name string) exported.ModuleAccountI
+	GetModuleAccount(ctx sdk.Context, name string) supply.ModuleAccountI
 
 	SendCoinsFromModuleToAccount(ctx sdk.Context, senderModule string, recipientAddr sdk.AccAddress, amt sdk.Coins) sdk.Error
 	SendCoinsFromAccountToModule(ctx sdk.Context, senderAddr sdk.AccAddress, recipientModule string, amt sdk.Coins) sdk.Error
 	SendCoinsFromModuleToModule(ctx sdk.Context, senderModule, recipientModule string, amt sdk.Coins) sdk.Error
 }
+
+type AccountKeeper interface {
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/expected_keepers.go]: comment

File: modules/asset/01-token/internal/types/genesis.go [hunk 15]
```diff
@@ -16,14 +30,25 @@ func NewGenesisState(params Params, tokens Tokens) GenesisState {
...
+	}
+	return data.Params.Validate()
+}
+
+func DefaultTokens() Tokens {
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/genesis.go]: comment

File: modules/asset/01-token/internal/types/genesis.go [hunk 16]
```diff
@@ -1,5 +1,19 @@
...
+)
+
+var DefaultToken = FungibleToken{
+	Symbol:        types.Iris,
+	Name:          "IRIS Network",
+	Scale:         18,
+	MinUnit:       types.IrisAtto,
+	InitialSupply: sdk.NewIntWithDecimal(20, 9),
+	Mintable:      true,
+}
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/genesis.go]: would it be better to use like this?
```
func DefaultToken() FungibleToken {
  return FungibleToken{...}
}
```

File: modules/asset/01-token/internal/types/token.go [hunk 17]
```diff
@@ -0,0 +1,264 @@
+package types
```
Comment (SegueII) [inline modules/asset/01-token/internal/types/token.go]: complete comments for this file

File: modules/asset/01-token/handler.go [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+			sdk.NewAttribute(AttributeKeyTokenSymbol, msg.Symbol),
+		),
+	})
+
+	return sdk.Result{
+		Events: ctx.EventManager().Events(),
+	}
+}
+
+// handleMsgMintToken handles MsgMintToken
```
Comment (SegueII) [inline modules/asset/01-token/handler.go]:
```suggestion
// HandleMsgMintToken handles MsgMintToken
```

File: modules/asset/keeper/keeper.go [hunk 19]
```diff
@@ -0,0 +1,23 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/params"
+	token "github.com/irisnet/irishub/modules/asset/01-token"
+)
+
+type Keeper struct {
```
Comment (SegueII) [inline modules/asset/keeper/keeper.go]: comments

File: modules/asset/codec.go [hunk 20]
```diff
@@ -0,0 +1,19 @@
+package asset
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	token "github.com/irisnet/irishub/modules/asset/01-token"
+)
+
```
Comment (SegueII) [inline modules/asset/codec.go]:
```suggestion
// ModuleCdc defines the module codec
```

File: modules/asset/codec.go [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+var ModuleCdc *codec.Codec
```
Comment (SegueII) [inline modules/asset/codec.go]:
```suggestion
// ModuleCdc defines the module codec
var ModuleCdc *codec.Codec
```

File: modules/asset/01-token/internal/types/genesis.go (hunk 16 above)
Comment (zhiqiang-bianjie) [reply modules/asset/01-token/internal/types/genesis.go]: no need

**Pull request merged**